ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,392 @@
+   * @param segment the {@link LogSegment} whose "next" segment is required.
+   * @return the {@link LogSegment} that is logically after the given {@code segment}.
+   */
+  LogSegment getNextSegment(LogSegment segment) {
+    String name = segment.getName();
+    if (!segmentsByName.containsKey(name)) {
+      throw new IllegalArgumentException("Invalid log segment name: " + name);
     }
-    int bytesWritten = fileChannel.write(buffer, currentWriteOffset.get());
-    currentWriteOffset.addAndGet(bytesWritten);
-    logger.trace("Log: {} bytes appended to the log from bytebuffer byteswritten : {}", file.getAbsolutePath(),
-        bytesWritten);
-    return bytesWritten;
+    Map.Entry<String, LogSegment> nextEntry = segmentsByName.higherEntry(name);
+    return nextEntry == null ? null : nextEntry.getValue();
   }
 
-  @Override
-  public void appendFrom(ReadableByteChannel channel, long size)
+  /**
+   * @param name the name of the segment required.
+   * @return a {@link LogSegment} with {@code name} if it exists.
+   */
+  LogSegment getSegment(String name) {
+    return segmentsByName.get(name);
+  }
+
+  /**
+   * @return the start offset of the log abstraction.
+   */
+  Offset getStartOffset() {
+    LogSegment segment = getFirstSegment();
+    return new Offset(segment.getName(), segment.getStartOffset());
+  }
+
+  /**
+   * @return the end offset of the log abstraction.
+   */
+  Offset getEndOffset() {
+    LogSegment segment = activeSegment;
+    return new Offset(segment.getName(), segment.getEndOffset());
+  }
+
+  /**
+   * @return the currently capacity used of the log abstraction. Includes "wasted" space at the end of
+   * {@link LogSegment} instances that are not fully filled.
+   */
+  long getUsedCapacity() {
+    return getDifference(getEndOffset(), new Offset(getFirstSegment().getName(), 0));
+  }
+
+  /**
+   * @return the number of valid segments starting from the first segment.
+   */
+  long getSegmentCount() {
+    return segmentsByName.size();
+  }
+
+  /**
+   * Gets the absolute difference in bytes between {@code o1} and {@code o2}. The difference returned also includes the
+   * sizes of log segment headers if the offsets are across segments.
+   * @param o1 the first {@link Offset}.
+   * @param o2 the second {@link Offset}.
+   * @return the difference between {@code o1} and {@code o2}. If {@code o1} > {@code o2}, the difference returned is
+   * positive, else, unless equal, it is negative
+   */
+  long getDifference(Offset o1, Offset o2) {
+    if (!segmentsByName.containsKey(o1.getName()) || !segmentsByName.containsKey(o2.getName())) {
+      throw new IllegalArgumentException("One of the log segments provided [" + o1.getName() + ", " + o2.getName() +
+          "] does not belong to this log");
+    }
+    LogSegment firstSegment = segmentsByName.get(o1.getName());
+    LogSegment secondSegment = segmentsByName.get(o2.getName());
```

Comment:
They are not start and end offsets because start and end gives the impression that start <= end (which is not the case here). I have avoided naming them start-end for that reason.